### PR TITLE
Async.debounce / Async.throttle Type Updates

### LINKED ISSUE #14188: Type for Async.debounce does not account for arguments
### Environment Information

- **Package version(s)**: 7.123.8
- **Browser and OS versions**: N/A

### Please provide a reproduction of the bug in a codepen:
It looks like the types for Async.debounce does not account for the arguments of the Function.

https://codesandbox.io/s/elastic-swartz-6vrpj?file=/src/App.tsx

Here is the raw TS also.
```
import { Async } from "office-ui-fabric-react/lib/Utilities";

function printAddOne(num: number): void {
  console.log(num + 1);
}

const async = new Async();
const debounced = async.debounce(printAddOne, 500);
const debounced2 = async.debounce<(n: number) => void>(printAddOne, 500);

debounced(300);
debounced2(300);
```

Here is the signature of debounce:
```
public debounce<T extends Function>(
    func: T,
    wait?: number,
    options?: {
      leading?: boolean;
      maxWait?: number;
      trailing?: boolean;
    },
  ): ICancelable<T> & (() => void) {
```

The issue is the last `() => void)`, which means the function can only ever be argument-less and void according to TypeScript.

#### Actual behavior:

Cannot call a Async.debounce function with an argument

#### Expected behavior:

No TypeScript error occurs when calling the debounced function with expected arguments.

### Priorities and help requested:

Are you willing to submit a PR to fix? Yes

Requested priority: Normal

Products/sites affected: N/A

### PR DESCRIPTION
**Pull request by SpencerLynn**
…clude a better inferred type for the throttled/debounced function

#### Pull request checklist

- [x] Addresses an existing issue: Fixes #14188
- [x] Include a change request file using `$ yarn change`

#### Description of changes

Updating the type definitions for the Async.debounce and Async.throttle to include more correct type inference of the throttled/debounced function.

#### Focus areas to test

Types around existing usages.

### CODE REVIEW

File: packages/utilities/etc/utilities.api.md [hunk 1]
```diff
@@ -64,10 +64,10 @@ export class Async {
     setImmediate(callback: () => void, targetElement?: Element | null): number;
     setInterval(callback: () => void, duration: number): number;
     setTimeout(callback: () => void, duration: number): number;
-    throttle<T extends Function>(func: T, wait?: number, options?: {
+    throttle<T extends (...args: any[]) => any>(func: T, wait?: number, options?: {
         leading?: boolean;
         trailing?: boolean;
-    }): T | (() => void);
+    }): T;
```
Comment (SpencerLynn) [inline packages/utilities/etc/utilities.api.md:70]: Are there any tests for this? I'd be interested in potentially adding tests that catch the compilation issue I ran across.

**Comment by SpencerLynn** [comment]
@xugao I am working on some unit tests for some of this, that hopefully would have caught the type issue (along with some other random testing).

I'll be pushing it soon!

File: change/@uifabric-utilities-2020-07-24-13-09-30-debounce-type-definition.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "Updating debounce and throttle type definitions for correctness to include a better inferred type for the throttled/debounced function",
+  "packageName": "@uifabric/utilities",
+  "email": "sllynn8907@gmail.com",
+  "dependentChangeType": "patch",
```
[the remark this answers is not in the document]
Comment (SpencerLynn) [reply change/@uifabric-utilities-2020-07-24-13-09-30-debounce-type-definition.json:6]: @xugao what do you think about this being a patch? Not sure where "types updated, but no functionality changed" falls into this category :)

**Comment by SpencerLynn** [comment]
@xugao, if you have time, please check out my latest push :) I added some tests. I also updated a few usages of `new Date().getTime()` to simply `Date.now()`, so I could more easily mock those responses to properly test `Async.debounce` and `Async.throttle`. Let me know if you see any issues. Thanks!

File: packages/utilities/src/Async.test.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    let async: Async;
+    let debouncedFn: ICancelable<typeof fn> & typeof fn;
+
+    beforeEach(() => {
+      jest.useFakeTimers();
+
+      callCount = 0;
+      async = new Async();
+      fn.mockClear();
```
Comment (xugao) [inline packages/utilities/src/Async.test.ts:21]: clear mock `afterEach`

File: change/@uifabric-utilities-2020-07-24-13-09-30-debounce-type-definition.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (xugao) [reply change/@uifabric-utilities-2020-07-24-13-09-30-debounce-type-definition.json:6]: this is correct :)

File: packages/utilities/src/Async.test.ts (hunk 3 above)
Comment (SpencerLynn) [reply packages/utilities/src/Async.test.ts]: Moved mock and `callCount` clearing to `afterEach`. Thanks!

**Comment by SpencerLynn** [comment]
@xugao Let me know if there is anything else I need to do on this PR :)

**Comment by dzearing** [comment]
lgtm :)

**Comment by SpencerLynn** [comment]
If anyone who has write permissions and approves of this PR wants to merge it, I'd be greatly appreciative :)

**Comment by SpencerLynn** [comment]
@xugao @dzearing Thanks for taking a look. I'm done with my changes and the PR is ready to be merged on my end. Let me know if you need anything else!

**Comment by SpencerLynn** [comment]
Sorry to ping again here, but is there anything I can do to help complete this PR, @xugao ?

**Comment by SpencerLynn** [comment]
@paulgildea, Sorry to ping you, but do you mind helping get some traction on this PR? :)

**Comment by SpencerLynn** [comment]
@khmakoto I'd love to get some more traction on this PR if you have some time :) It's been reviewed, but I am unable to merge it.

**Comment by khmakoto** [comment]
@SpencerLynn for sure, let me merge it for you!

**Pull request merged**